crates/bevy_ecs/src/world/mod.rs

```diff
@@ -1293,29 +1419,52 @@ impl World {
         &self,
         component_id: ComponentId,
     ) -> Option<Mut<'_, R>> {
-        self.validate_non_send_access::<R>();
-        self.get_resource_unchecked_mut_with_id(component_id)
+        let (ptr, ticks) = self
+            .storages
+            .non_send_resources
+            .get(component_id)?
+            .get_with_ticks()?;
+        Some(Mut {
+            value: ptr.assert_unique().deref_mut(),
+            ticks: Ticks {
+                added: ticks.added.deref_mut(),
+                changed: ticks.changed.deref_mut(),
+                last_change_tick: self.last_change_tick(),
+                change_tick: self.read_change_tick(),
+            },
+        })
     }
 
     /// Inserts a new resource with the given `value`. Will replace the value if it already existed.
     ///
     /// **You should prefer to use the typed API [`World::insert_resource`] where possible and only
     /// use this in cases where the actual types are not known at compile time.**
     ///
+    /// # Panics
+    /// If `is_send` is false, and a value is already present, this function will panic if called
+    /// from a thread that the original value was inserted from.
+    ///
     /// # Safety
     /// The value referenced by `value` must be valid for the given [`ComponentId`] of this world
-    /// `component_id` must exist in this [`World`]
+    /// `component_id` must exist in this [`World`]. `is_send` must correspond to whether the resource
+    /// adheres to invariants of `Send`.
     #[inline]
     pub unsafe fn insert_resource_by_id(
         &mut self,
         component_id: ComponentId,
+        is_send: bool,
```

Comment:
All the call sites of this in bevy seem to just be using `true`/`false` literals so it probably makes sense to split this like most of the other resource apis by having both `insert_resource_by_id` and `insert_non_send_by_id` :thinking: